lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs

```diff
@@ -0,0 +1,183 @@
+    ( StandardCrypto
+    )
+import Cardano.Wallet.Read.Address
+    ( toShortByteString
+    )
+import Codec.Binary.Bech32
+    ( DataPart
+    , HumanReadablePart
+    , dataPartFromBytes
+    , dataPartToBytes
+    , decodeLenient
+    )
+import Control.Arrow
+    ( ArrowChoice (..)
+    )
+import Control.Monad
+    ( (>=>)
+    )
+import Control.Monad.State.Strict
+    ( evalStateT
+    )
+import Data.ByteString
+    ( ByteString
+    )
+import Data.ByteString.Base58
+    ( bitcoinAlphabet
+    , decodeBase58
+    , encodeBase58
+    )
+import Data.Text
+    ( Text
+    )
+
+import qualified Cardano.Ledger.Address as SH
+import qualified Cardano.Ledger.Address as SL
+import qualified Cardano.Ledger.BaseTypes as SL
+import qualified Codec.Binary.Bech32 as Bech32
+import qualified Codec.Binary.Bech32.TH as Bech32
+import qualified Data.ByteString.Short as B8
+import qualified Data.Text.Encoding as T
+
+data AddressFlavor a
+    = Bootstrap
+        {addressFlavor :: a}
+    | Shelley
+        {addressFlavor :: a}
+    deriving (Eq, Show)
+
+-- | Errors that can occur when decoding an 'Address'.
+data DecodingError
+    = InvalidBech32Encoding Bech32.DecodingError
+    | InvalidBase58Encoding
+    | InvalidHumanReadablePart HumanReadablePart
+    | InvalidDataPart DataPart
+    | InvalidNetwork
+    | AddressFlavorMismatch
+    | AddressDecodingError String
+    | AddressNetworkMismatch
+    deriving (Eq, Show)
+
+humanPart :: NetworkTag -> HumanReadablePart
+humanPart = \case
+    MainnetTag -> [Bech32.humanReadablePart|addr|]
+    TestnetTag -> [Bech32.humanReadablePart|addr_test|]
+
+decodeBase58Address
+    :: ByteString -> Either DecodingError (AddressFlavor ByteString)
+decodeBase58Address =
+    fmap Bootstrap
+        . maybe (Left InvalidBase58Encoding) Right
+        . decodeBase58 bitcoinAlphabet
+
+decodeBech32Address
+    :: NetworkTag -> Text -> Either DecodingError (AddressFlavor ByteString)
+decodeBech32Address tag bech32 = do
+    (hrp, dataPart) <- left InvalidBech32Encoding $ decodeLenient bech32
+    if hrp /= humanPart tag
+        then Left $ InvalidHumanReadablePart hrp
+        else case dataPartToBytes dataPart of
+            Nothing -> Left $ InvalidDataPart dataPart
+            Just bytes -> pure $ Shelley bytes
+
+decodeHumanAddress
+    :: NetworkTag -> Text -> Either DecodingError (AddressFlavor ByteString)
+decodeHumanAddress tag t =
+    decodeBech32Address tag t
+        <> decodeBase58Address (T.encodeUtf8 t)
+
+newtype CatchFail a = CatchFail {runCatchFail :: Either String a}
+    deriving (Functor, Applicative, Monad)
+
+instance MonadFail CatchFail where
+    fail = CatchFail . Left
+
+ledgerAddressFlavor :: SL.Addr c -> AddressFlavor ()
+ledgerAddressFlavor (SL.AddrBootstrap _) = Bootstrap ()
+ledgerAddressFlavor _ = Shelley ()
+
+ledgerAddressNetwworkTag :: SL.Addr c -> NetworkTag
+ledgerAddressNetwworkTag addr = case SL.getNetwork addr of
+    SL.Testnet -> TestnetTag
```

Comment:
we are sure we want isomorphic and so similar type to the ledger here?

Reply:
I do not have control on this, but happy to remove this when it's removed from the deposit agda code.